lib/imagemounter/utils/list_developer_image.js

```diff
@@ -0,0 +1,182 @@
+        if (nextItem.type !== 'tree') {
+            throw Error(`${entry} under under ${fileList.join('/')} in under ${repoUrl} is expected to be a tree, got ${nextItem.type} instead.`);
+        }
+        fileList.push(entry);
+        curUrl = nextItem.url;
+    }
+    return ret;
+}
+
+/**
+ * Find `DeveloperDiskImage.dmg` recursively under folder and subfolder.
+ * @param {string} entry current folder
+ * @returns {string | undefined} parent folder of `DeveloperDiskImage.dmg` or `undefined` if no such file exists
+ */
+function findDeveloperImageFromDirectory(entry) {
+    const fileList = readdirSync(entry);
+    for (const subEntry of fileList) {
+        const fullPath = joinPath(entry, subEntry);
+        const statResult = lstatSync(fullPath);
+        if (subEntry === DEVELOPER_IMAGE_FILE_NAME && statResult.isFile()) {
+            return entry;
+        }
+        if (statResult.isDirectory()) {
+            const subFolderResult = findDeveloperImageFromDirectory(fullPath);
+            if (subFolderResult) {
+                return subFolderResult;
+            }
+        }
+    }
+    return undefined;
+}
+
+/**
+ * @typedef {Object} ImagePath
+ * @property {string} developerImage
+ * @property {string} developerImageSignature
+ */
+
+/**
+ * Find developer image for certain version. If developer image does not exists,
+ * this will try to find and download developer image, unzip to `~/.appium/iOSSupport/`
+ * @param {string} version full version of iOS device
+ * @returns {ImagePath}
+ * @throws If developer image is not found, or error while downloading or unzipping.
+ */
+async function findDeveloperImage(version) {
+    const finalVersion = version.split('.').splice(0, 2).join('.');
+    const fileName = `${finalVersion}.zip`;
+    const fullDownloadPath = joinPath(DEFAULT_IMAGE_DIR, fileName);
+    if (!existsSync(DEFAULT_IMAGE_DIR)) {
+        mkdirSync(DEFAULT_IMAGE_DIR, { recursive: true });
+    }
+    if (!existsSync(fullDownloadPath)) {
+        await searchAndDownloadDeveloperImageFromGithub(finalVersion, fullDownloadPath);
+    }
+    const decompressPath = joinPath(DEFAULT_IMAGE_DIR, finalVersion);
+    let developerImageParentFolder = null;
+    if (existsSync(decompressPath)) {
+        developerImageParentFolder = findDeveloperImageFromDirectory(decompressPath);
+    }
+    if (!developerImageParentFolder) {
+        await zip.extractAllTo(fullDownloadPath, decompressPath);
+        developerImageParentFolder = findDeveloperImageFromDirectory(decompressPath);
+    }
+    if (!developerImageParentFolder) {
+        throw Error(`Unable to find unzipped developer image in ${decompressPath}`);
+    }
+    return {
+        developerImage: joinPath(developerImageParentFolder, DEVELOPER_IMAGE_FILE_NAME),
+        developerImageSignature: joinPath(developerImageParentFolder, DEVELOPER_IMAGE_SIGNATURE_FILE_NAME)
+    };
+}
+
+async function searchAndDownloadDeveloperImageFromGithub(finalVersion, fullDownloadPath) {
```

Comment:
this logic could be tested (and probably must be tested)

Reply:
I've tested on my own devices on both macOS and ubuntu

Reply:
this is great. Although something could easily be broken if any changes are added into this logic. That is why it would be much safer if we had this logic present as part of integration tests

Reply:
agree; we need tests for this

Reply:
Could we download a real image file from a real repo as a test? I'm not sure if this could be processed on github actions

Reply:
Especially when it comes to local file operations such as decompression, etc.

Reply:
> Could we download a real image file from a real repo as a test? I'm not sure if this could be processed on github actions

why not? I don't see any issues with that